### PR TITLE
YJIT: fix 32 and 16 bit register store

### PR DESCRIPTION
**Pull request by jemmaissroff**
Co-Authored-By: Takashi Kokubun <takashikkbn@gmail.com>

This will unblock #6767. Confirmed by cherry-picking these changes onto #6767 in draft PR (#6839)

### CODE REVIEW

File: yjit/src/backend/arm64/mod.rs [hunk 1]
```diff
@@ -153,7 +153,9 @@ impl Assembler
                         out_opnd
                     }
                 },
-                _ => asm.load(opnd)
+                _ => {
+                    asm.load(opnd)
+                }
```
Comment (k0kubun) [inline yjit/src/backend/arm64/mod.rs:158]:
```suggestion
                _ => asm.load(opnd)
```

It was an unneeded diff made during debugging

File: yjit/src/backend/arm64/mod.rs [hunk 2]
```diff
@@ -821,7 +827,11 @@ impl Assembler
                     // the Arm64 assembler works, the register that is going to
                     // be stored is first and the address is second. However in
                     // our IR we have the address first and the register second.
-                    stur(cb, src.into(), dest.into());
+                    match dest.rm_num_bits() {
+                        64 => stur(cb, src.into(), dest.into()),
+                        32 => sturh(cb, src.into(), dest.into()),
+                        _ => panic!("unexpected dest num_bits: {}, src: {:#?}, dest: {:#?}", dest.rm_num_bits(), src, dest),
```
Comment (k0kubun) [inline yjit/src/backend/arm64/mod.rs:833]: We could reuse the result of `rm_num_bits()`

```suggestion
                        num_bits => panic!("unexpected dest num_bits: {} (src: {:#?}, dest: {:#?})", num_bits, src, dest),
```

File: yjit/src/backend/arm64/mod.rs [hunk 3]
```diff
@@ -501,6 +501,12 @@ impl Assembler
                     // than 9 bits long. If it's longer, we need to load the
                     // memory address into a register first.
                     let opnd0 = split_memory_address(asm, dest);
+                    let opnd0 = match (opnd0, src) {
+                        (Opnd::Mem(dest), Opnd::UImm(src)) if dest.num_bits == 16 && src < 65536 => {
```
Comment (maximecb) [inline yjit/src/backend/arm64/mod.rs:505]:
```suggestion
                        (Opnd::Mem(dest), Opnd::UImm(src)) if dest.num_bits == 16 && src < u16::MAX => {
```

File: yjit/src/asm/arm64/inst/load_store.rs [hunk 4]
```diff
@@ -118,6 +118,12 @@ impl LoadStore {
     pub fn stur(rt: u8, rn: u8, imm9: i16, num_bits: u8) -> Self {
         Self { rt, rn, idx: Index::None, imm9, opc: Opc::STR, size: num_bits.into() }
     }
+
+    /// STURH (store register, halfword, unscaled)
+    /// https://developer.arm.com/documentation/ddi0596/2021-12/Base-Instructions/STUR--Store-Register--unscaled--?lang=en
```
Comment (XrXr) [inline yjit/src/asm/arm64/inst/load_store.rs:123]:
```suggestion
    /// https://developer.arm.com/documentation/ddi0596/2021-12/Base-Instructions/STURH--Store-Register-Halfword--unscaled--?lang=en
```
link wasn't changed.

File: yjit/src/backend/arm64/mod.rs [hunk 5]
```diff
@@ -821,7 +827,11 @@ impl Assembler
                     // the Arm64 assembler works, the register that is going to
                     // be stored is first and the address is second. However in
                     // our IR we have the address first and the register second.
-                    stur(cb, src.into(), dest.into());
+                    match dest.rm_num_bits() {
+                        64 => stur(cb, src.into(), dest.into()),
+                        32 => sturh(cb, src.into(), dest.into()),
```
Comment (XrXr) [inline yjit/src/backend/arm64/mod.rs:832]: It's odd to that we use a 16 bit store when the size is 32 bit. Does it store the whole word properly when the destination is 32 bit?

Comment (k0kubun) [reply yjit/src/backend/arm64/mod.rs:832]: Oh I didn't realize "halfword" means 16 bits. I guess we should just use `sturh` for 16 and `stur` for 64/32 then.

**Pull request merged**